bundles/org.openhab.binding.qolsysiq/src/main/java/org/openhab/binding/qolsysiq/internal/client/QolsysiqClient.java

```diff
@@ -0,0 +1,379 @@
+import org.openhab.binding.qolsysiq.internal.client.dto.event.ZoneActiveEvent;
+import org.openhab.binding.qolsysiq.internal.client.dto.event.ZoneEventType;
+import org.openhab.binding.qolsysiq.internal.client.dto.event.ZoneUpdateEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * A client that can communicate with a Qolsys IQ Panel
+ *
+ * @author Dan Cunningham - Initial contribution
+ */
+@NonNullByDefault
+public class QolsysiqClient {
+    private static final String MESSAGE_ACK = "ACK";
+    private final Logger logger = LoggerFactory.getLogger(QolsysiqClient.class);
+    private final Gson gson = new GsonBuilder().registerTypeAdapter(Event.class, new EventDeserializer())
+            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
+    private List<QolsysIQClientListener> listeners = Collections.synchronizedList(new ArrayList<>());
+    private @Nullable SSLSocket socket;
+    private @Nullable BufferedReader reader;
+    private @Nullable BufferedWriter writer;
+    private @Nullable Thread readerThread;
+    private @Nullable ScheduledFuture<?> heartBeatFuture;
+    private ScheduledExecutorService scheduler;
+    private Object writeLock = new Object();
+    private long lastResponseTime;
+    private boolean hasACK = false;
+    private boolean connected;
+    private String host;
+    private int port;
+    private int heartbeatSeconds;
+
+    /**
+     * Creates a new QolsysiqClient
+     *
+     * @param host
+     * @param port
+     * @param heartbeatSeconds
+     * @param scheduler for the heart beat task
+     */
+    public QolsysiqClient(String host, int port, int heartbeatSeconds, ScheduledExecutorService scheduler) {
+        this.host = host;
+        this.port = port;
+        this.heartbeatSeconds = heartbeatSeconds;
+        this.scheduler = scheduler;
+    }
+
+    /**
+     * Connects to the panel
+     *
+     * @throws IOException
+     */
+    public synchronized void connect() throws IOException {
+        logger.debug("connect");
+        if (connected) {
+            logger.debug("connect: already connected, ignoring");
+            return;
+        }
+        SSLSocketFactory sslsocketfactory;
+        try {
+            SSLContext sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(null, acceptAlltrustManagers(), null);
+            sslsocketfactory = sslContext.getSocketFactory();
+        } catch (KeyManagementException | NoSuchAlgorithmException e) {
+            throw new IOException(e);
+        }
```

Comment:
> Maybe it could be initialized in the constructor

So i had it this way originally,  which threw a runtime exception if the `SSLSocketFactory` could not be created, unfortunately this caused a checkstyle warning , so i moved it here to quiet it. I'll move it back, unless you have another thought? 

>or even globally in the thing handler factory 

I was trying to keep the client code as contained as possible

Reply:
In practice, there is no reason that this portion of code throws an exception, do you agree ?
Maybe it will be more secure that in the your thing handler method connect(), you put the call to the client inside your try/catch. WDYT ?

Reply:
Yeah, I for some reason had a huge hole in my memory and forgot constructors can throw checked exceptions 🤦